File: piper/extensions/chrome/src/engine/statements/controls_repeat_ext.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        this.loopFlow = null
        let pointer = statement.pointer
        let times = await this.executeStatement(stringify(pointer.times)) || 0
        times = parseInt(times)
        let loopBlock = stringify(pointer.loopBlock).trim()
        if (loopBlock !== "") {
            for (let i = 0; i < times; i++) {
                await this.execute(loopBlock)
                if (this.isTerminated) break
                if (this.loopFlow === "BREAK") {
                    this.loopFlow = null
                    break
                } else if (this.loopFlow === "CONTINUE") {
                    this.loopFlow = null
                    continue
                }

                if (this.isFunctionReturn != null) break
            }
        }
    } catch (e) {
        onError(e)
    }
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/controls_repeat_ext.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeControlsRepeatExt(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_controls_repeat_ext")
        loopFlow = null
        val pointer = statement.getJSONObject("pointer")
        val times = executeStatement(pointer["times"].toString())?.toString()?.toInt() ?: 0
        val loopBlock = pointer["loopBlock"].toString().trim()
        if (loopBlock != "") {
            for (i in 1..times) {
                execute(loopBlock)
                if (isTerminated) break
                if (loopFlow == "BREAK") {
                    loopFlow = null
                    break
                } else if (loopFlow == "CONTINUE") {
                    loopFlow = null
                    continue
                }

                if (isFunctionReturn != null) break
            }
        }
    } catch (e: Exception) {
        onError(e)
    }
}